File: C++/addInList.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* addInList(ListNode* head1, ListNode* head2) {
        ListNode *reversed_head1 = reverse(head1);
        ListNode *reversed_head2 = reverse(head2);
        ListNode *dummy = new ListNode(0);
        ListNode *pre = dummy;
        int carry = 0;
        while (reversed_head1 || reversed_head2 || carry) {
            int sum = carry;
            if (reversed_head1) {
                sum += reversed_head1->val;
                reversed_head1 = reversed_head1->next;
            }
            if (reversed_head2) {
                sum += reversed_head2->val;
                reversed_head2 = reversed_head2->next;
            }
            carry = sum / 10;
            pre->next = new ListNode(sum % 10);
            pre = pre->next;
        }
        return reverse(dummy->next);
    }

private:
    ListNode *reverse(ListNode* head) {
        ListNode *prev = nullptr;
        ListNode *current = head;
        while (current) {
            ListNode *next = current->next;
            current->next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }
};

void deleteList(ListNode* &head) {
    while (head) {
        ListNode *temp = head;
        head = head->next;
        delete temp;
    }
}

int main() {
    ListNode* head1 = new ListNode(9);
    head1->next = new ListNode(3);
    head1->next->next = new ListNode(7);
    ListNode* head2 = new ListNode(6);
    head2->next = new ListNode(3);
    Solution solution;

    ListNode* addHead = solution.addInList(head1, head2);
    ListNode* current = addHead; 
    while (current) {
        std::cout << current->val << " ";
        current = current->next;
    }
    std::cout << std::endl;

    deleteList(addHead); 

    return 0;
}

File: python/addInList.py
from os import pread

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addInList(self, head1: ListNode, head2: ListNode) -> ListNode:
        if head1 is None or head2 is None:
            return head2 if head1 is None else head1

        reversed_head1 = self.reverse(head1)
        reversed_head2 = self.reverse(head2)

        flag = 0
        pre = None

        while reversed_head1 is not None or reversed_head2 is not None:
            num = flag

            if reversed_head1 is not None:
                num += reversed_head1.val
                reversed_head1 = reversed_head1.next

            if reversed_head2 is not None:
                num += reversed_head2.val
                reversed_head2 = reversed_head2.next

            flag = num // 10
            current = ListNode(num % 10)
            current.next = pre
            pre = current

        if flag:
            new_node = ListNode(flag)
            new_node.next = pre
            pre = new_node

        return pre

    def reverse(self, head: ListNode) -> ListNode:
        current = head
        pre = None

        while current is not None:
            next_node = current.next
            current.next = pre
            pre = current
            current = next_node

        return pre

head1=ListNode(9)
head1.next=ListNode(3)
head1.next.next=ListNode(7)
head2=ListNode(6)
head2.next=ListNode(3)
solution = Solution()
add_head=solution.addInList(head1,head2)
while add_head:
    print(add_head.val,end='')
    add_head=add_head.next
print()
    